File: cookbook/php/Segments/createSegmentStatic.php
<?php

//Ici, renseignez la xKey et les parametres personnalises du message
$xKey = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
$segmentId = 0123;	//Id du segment a modifier ('null' si le segment est a creer)
$unicity = 'test@test.com'; //L'adresse mail de la cible a mettre dans le segment ('null' pour ne rien ajouter au segment)

$type = 'static';	//Code pour creer un segment Statique
$name = 'SegmentStatic (php)';	//Nom du segment
$description = 'SegmentStatic (php)';	//Description du segment
$expiration = '2016-01-08T12:11:00Z';	//Date d'expiration du segment
$isTest = true;	//Segment de test : oui = 'true' / non = 'false'

//On trouve l'adresse pour la requete
$url = 'http://v8.mailperformance.com/segments/' . $segmentId;


//Creation du Json du message
$arr = array(
	'type' => $type,
	'name' => $name,
	'description' => $description,
	'expiration' => $expiration,
	'isTest' => $isTest);
		

//On affiche le message
$message = json_encode($arr);
echo $message . "\n";

//Connexion
if ($segmentId != null)
	$con = connect($url, $xKey, $message, 'PUT');
else
	$con = connect($url, $xKey, $message, 'POST');

$result = $con['result'];
$info = $con['info'];
$req = $con['req'];

if ($info['http_code'] != 200)
{
	echo 'Error : ' . $info['http_code'];
}
else
{
	//Le segment a bien ete cree
	echo 'The segment static  : ' . $name . " is created / update.\n\n";
	
	if ($unicity != null && $segmentId != null)
	{
		//Nouvelle url
		$url = 'http://v8.mailperformance.com/targets?unicity=' . $unicity;
	
		//Connexion
		$con = connect($url, $xKey, null, 'GET');
	
		$result = $con['result'];
		$info = $con['info'];
		$req = $con['req'];
	
	
		if ($info['http_code'] != 200)
		{
			echo 'Error : ' . $info['http_code'];
		}
		else
		{
			$tab = json_decode($result, TRUE);
			$idTarget = $tab['id'];
	
			//Nouvelle url
			$url = 'http://v8.mailperformance.com/targets/' . $idTarget . '/segments/' . $segmentId;
	
			//Connexion
			$con = connect($url, $xKey, null, 'POST');
		
			$result = $con['result'];
			$info = $con['info'];
			$req = $con['req'];
	
			if ($info['http_code'] != 204)
				echo 'Error : ' . $info['http_code'];
			else
				echo 'The target : ' . $unicity . ' is add to the segment : ' . $name;
		}
	}
}
curl_close($req);



//Fonctions -----



//Utilisation de cURL pour remplir les requetes
function startCurlInit($url)
{
	$init = curl_init();
	curl_setopt($init, CURLOPT_URL, $url);
	curl_setopt($init, CURLOPT_RETURNTRANSFER, true);
	return ($init);
}

//Fonction de connexion
function connect($url, $xKey, $message, $method)
{
	//On remplit la requete
	$req = startCurlInit($url);
	curl_setopt($req, CURLOPT_CUSTOMREQUEST, $method);
		
	//Mise en place du xKey et des options
	curl_setopt($req, CURLOPT_HTTPHEADER, array(
	'X-Key: ' . $xKey,
	'Content-Type: application/json',
	'Content-Length: ' . strlen($message)));
	curl_setopt($req, CURLOPT_POSTFIELDS, $message);

	//Execution de la requete
	$result = curl_exec($req);

	//Verification des reponses
	$info = curl_getinfo($req);
	
	return (array('result' => $result, 'info' => $info, 'req' => $req));
}

?>

File: cookbook/Csharp/Segments/createSegmentStatic.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.IO;
using System.Net;

namespace createSegmentStatic
{
	class Program
	{
		static void Main(string[] args)
		{
			//Ici, renseignez la xKey et les parametres personnalises
			String xKey = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
			String segmentId = "0123";	//Id du segment a modifier ('null' si le segment doit etre cree)
			String unicity = "test@test.com"; //L'adresse mail de la cible a mettre dans le segment ('null' pour ne rien ajouter au segment)

			String type = "static";	//Code pour creer un segment Statique
			String name = "SegmentStatic (Csharp)";	//Nom du segment
			String description = "SegmentStatic (Csharp)";	//Description du segment
			String expiration = "2016-01-08T12:11:00Z";	//Date d'expiration du segment
			Boolean isTest = true;	//Le test est un segment de test : oui = 'true' / non = 'false'


			//On trouve l'adresse pour la requete
			String url = "http://v8.mailperformance.com/segments/";
			if (segmentId != null)
			{
				url = "http://v8.mailperformance.com/segments/" + segmentId;
			}

			//Creation du Json du message
			JObject jsonData = new JObject();
			jsonData.Add("type", type);
			jsonData.Add("name", name);
			jsonData.Add("description", description);
			jsonData.Add("expiration", expiration);
			jsonData.Add("isTest", isTest);

			//On affiche le message
			Console.Write(jsonData + "\n");


			//Lancement de la connexion pour remplir la requete
			Object[] connection = null;

			if (segmentId != null)
			{
				connection = allConnection(url, xKey, jsonData, "PUT");
			}
			else
			{
				connection = allConnection(url, xKey, jsonData, "POST");
			}

			int response = (int)connection[0];
			HttpWebRequest con = (HttpWebRequest)connection[1];
			HttpWebResponse httpResponse = (HttpWebResponse)connection[2];
			String responseString = (String)connection[3];

			//Verification des reponses
			if (response != 200)
			{
				//Affichage de l'erreur
				Console.Write("\nError : " + response);
			}
			else
			{
				//Le segment a bien ete cree
				Console.Write("The segment static  : " + name + " is created / update.\n\n");

				if (unicity != null && segmentId != null)
				{
					//Nouvelle url
					url = "http://v8.mailperformance.com/targets?unicity=" + unicity;

					//Connexion
					connection = allConnection(url, xKey, null, "GET");

					response = (int)connection[0];
					con = (HttpWebRequest)connection[1];
					httpResponse = (HttpWebResponse)connection[2];
					responseString = (String)connection[3];

					//Verification des reponses
					if (response != 200)
					{
						//Affichage de l'erreur
						Console.Write("\nError : " + response);
					}
					else
					{
						//On prend le idTarget
						dynamic responseJson = JsonConvert.DeserializeObject(responseString);
						string idTarget = responseJson.id;

						//Nouvelle url
						url = "http://v8.mailperformance.com/targets/" + idTarget + "/segments/" + segmentId;

						//Connexion
						connection = allConnection(url, xKey, null, "POST");

						response = (int)connection[0];
						con = (HttpWebRequest)connection[1];
						httpResponse = (HttpWebResponse)connection[2];
						responseString = (String)connection[3];

						//Verification des reponses
						if (response != 200)
						{
							//Affichage de l'erreur
							Console.Write("\nError : " + response);
						}
						else
						{
							Console.Write("The target : " + unicity + " is add to the segment : " + name);
						}
					}
				}
			}
			//Attente de lecture (optionnel)
			Console.ReadLine();

			httpResponse.Close();
		}



		//Fonctions ----



		//Fonction de connexion
		static HttpWebRequest Connect(string url, string xKey, string method)
		{
			//Lancement de la connexion pour remplir la requete
			HttpWebRequest con = (HttpWebRequest)WebRequest.Create(url);
			con.Method = method;

			//Mise en place du xKey et des options
			con.Headers.Add("X-Key", xKey);
			con.ContentType = "application/json";

			return (con);
		}

		//Fonction de connexion et envoie des informations
		static Object[] allConnection(String url, String xKey, JObject jsonMessage, String method)
		{
			HttpWebRequest con = Connect(url, xKey, method);
			con.ContentLength = 0;
			if (jsonMessage != null)
			{
				con.ContentLength = jsonMessage.ToString().Length;
				var streamWriter = new StreamWriter(con.GetRequestStream());
				streamWriter.Write(jsonMessage);
				streamWriter.Flush();
				streamWriter.Close();
			}

			//Test de l'envoie
			HttpWebResponse httpResponse = null;
			int response = 0;
			string responseString = null;
			try
			{
				httpResponse = (HttpWebResponse)con.GetResponse();
				response = 200;

				//Lecture des donnees
				StreamReader reader = new StreamReader(con.GetResponse().GetResponseStream());
				responseString = reader.ReadToEnd();
				httpResponse.Close();
				reader.Close();
			}
			//Reception du signal 
			catch (WebException ex)
			{
				if (ex.Status == WebExceptionStatus.ProtocolError)
				{
					httpResponse = (HttpWebResponse)ex.Response;
					response = (int)httpResponse.StatusCode;
				}
			}
			Object[] result = { response, con, httpResponse, responseString };
			return (result);
		}
	}
}
